{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "1e288650",
   "metadata": {},
   "source": [
    "# BME 590 - Workshop 5 - Interfacing with Peripherals and Synthetic Data\n",
    "**Professor:** Emma Chory, Ph.D.\n",
    "\n",
    "**Authors:** \n",
    "Rick Wierenga, Joe Laforet, Stefan Golas, Ben Perry\n",
    "\n",
    "---\n",
    "\n",
    "### Usage Note\n",
    "**Reminder** - You should be running this notebook **locally** on **VS Code** not navigating it through **GitHub**.\n",
    "\n",
    "**Reminder 2** - Remember to run `git pull` before copying this notebook so you get the most recent class updates. See [Section 6 of the class README](https://github.com/chory-lab/bme590-fall-2025#step-6-updating-to-the-latest-version)\n",
    "\n",
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dd9b0dbd",
   "metadata": {},
   "source": [
    "### Types of Peripherals\n",
    "\n",
    "Peripherals are machines that are wired into a liquid handling set-up. They can really be anything, such as:\n",
    "\n",
    "- Robotic arms\n",
    "\n",
    "- Centrifuges\n",
    "\n",
    "- Heater-Shakers\n",
    "\n",
    "- Plate Sealers\n",
    "\n",
    "- PCR Machines\n",
    "\n",
    "- Microscopes\n",
    "\n",
    "- Plate Readers\n",
    "\n",
    "- Storage\n",
    "\n",
    "- Refrigeration Units\n",
    "\n",
    "You have actually already interfaced with one of these peripherals, the robotic gripper. Unlike the gripper, however, many peripherals **generate data** in the form of absorbance values, fluorescence values, microscope images, quantification cycle counts, chromatograns, and more.\n",
    "\n",
    "While only some of these devices are explictly controllable via PLR, many can be controlled from an external library or set of customized backends that convert high level Python commands to firmware commands.\n",
    "\n",
    "Like workshop 3, we will **simulate** using these devices through Python classes. We are particularly interested in the ability to **simulate data**.\n",
    "\n",
    "Being able to simulate data is critical in laboratory robotics methods because:\n",
    "\n",
    "- It can be used to fully simulate a workflow **prior** to purchasing expensive equipment or making critcial errors.\n",
    "\n",
    "- It can be used to test the **extreme limits** of a workflow, such as when an assay produces unexpected data, null values, or maxes out.\n",
    "\n",
    "- It enables testing of **data processing pipelines** which may or may not include **AI methods** to analyze images and other results.\n",
    "\n",
    "To make the Peripheral accessible by our code, we will make a dummy class. Use the following code as inspiration should you wish to make a Peripheral for a different machine. The general idea is that the Peripheral is an object that produces formatted data. Most times, you will have to then perform some data processing to format the data correctly, and then apply a processing pipeline (like making a line of best fit for protein concentration or image analysis pipeline to get cell confluency), as most relevant quantities aren't directly measurable."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6aa9fc87",
   "metadata": {},
   "source": [
    "### Plate Reader Example\n",
    "A plate reader (also called a microplate reader or microplate spectrophotometer) is a laboratory instrument used to detect and measure biological, chemical, or physical reactions that occur in microplates. Some common modalities of measurement are:\n",
    "\n",
    "- **Absorbance:** Measures how much light is absorbed by a sample (common in enzyme assays and cell growth measurements).\n",
    "\n",
    "- **Fluorescence:** Detects light emitted by fluorescent molecules (used in DNA/RNA or protein quantification).\n",
    "\n",
    "- **Luminescence:** Measures light produced by chemical reactions (used in ATP or reporter gene assays).\n",
    "\n",
    "Let's showcase how we would simulate the data produced by a plate reader using a Python class.\n",
    "\n",
    "First, let's import our libraries.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "07001367",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from PIL import Image, ImageDraw\n",
    "\n",
    "from pylabrobot.liquid_handling import LiquidHandler\n",
    "from pylabrobot.liquid_handling.backends.backend import LiquidHandlerBackend\n",
    "from pylabrobot.liquid_handling.backends import LiquidHandlerChatterboxBackend\n",
    "from pylabrobot.resources import Plate, corning_96_wellplate_360ul_flat, OTDeck, Deck, opentrons_96_tiprack_300ul, axygen_1_reservoir_90ml, set_tip_tracking, set_volume_tracking, set_cross_contamination_tracking\n",
    "from pylabrobot.visualizer.visualizer import Visualizer"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ad133e80",
   "metadata": {},
   "source": [
    "Enable our trackers to throw errors and run visualizer correctly"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f9972539",
   "metadata": {},
   "outputs": [],
   "source": [
    "set_tip_tracking(enabled = True)\n",
    "set_volume_tracking(enabled = True)\n",
    "set_cross_contamination_tracking(enabled = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9be00cf3",
   "metadata": {},
   "source": [
    "Create our standard visualize deck function"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "65bd1b7e",
   "metadata": {},
   "outputs": [],
   "source": [
    "async def visualize_deck(deck: Deck,\n",
    "                         backend: LiquidHandlerBackend):\n",
    "    try:\n",
    "        lh = LiquidHandler(backend=backend, deck=deck)\n",
    "        vis = Visualizer(resource = lh)\n",
    "        await lh.setup()\n",
    "        await vis.setup()\n",
    "        return lh\n",
    "    except Exception as e:\n",
    "        print(f\"Error! Got excpetion: {e}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f78501a8",
   "metadata": {},
   "source": [
    "Now let's define a `PlateReader` class."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1a536d3b",
   "metadata": {},
   "outputs": [],
   "source": [
    "class PlateReader:\n",
    "\n",
    "    # define the constructor which can take in a place to save the absorbance data\n",
    "    def __init__(self,\n",
    "                 save_folder_path: str | os.PathLike):\n",
    "        self.save_folder = save_folder_path\n",
    "\n",
    "\n",
    "    # create a function to simulate reading a 96-well plate and saving the absorbance values to file.\n",
    "    def read_plate(self,\n",
    "                   plate: Plate, # for now, this argument doesn't do anything since we are hard-coding a 96 well plate.\n",
    "                   filename: str | os.PathLike = \"workshop_5_data.csv\"):\n",
    "\n",
    "        # list to save data\n",
    "        data = []\n",
    "\n",
    "        # for each row and column in the data, create psuedo-data from normal distribution\n",
    "        # centered at loc and with standard deviation of scale\n",
    "        for row in range(8):\n",
    "            row_data = []\n",
    "            for col in range(12):\n",
    "                absorbance_value = np.random.normal(loc=0.5,\n",
    "                                                    scale=0.1)\n",
    "                row_data.append(absorbance_value)\n",
    "            data.append(row_data)\n",
    "\n",
    "        # Write all generated data to a csv file\n",
    "        df = pd.DataFrame(data,\n",
    "                          index = [chr(65 + i) for i in range(8)],\n",
    "                          columns=[str(i + 1) for i in range(12)])\n",
    "        save_path = os.path.join(self.save_folder, filename)\n",
    "        df.to_csv(save_path, index=True, header=True)\n",
    "        print(f\"Wrote data to {save_path}!\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4e2ef507",
   "metadata": {},
   "source": [
    "Now let's initialize our plate reader"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "55eb1351",
   "metadata": {},
   "outputs": [],
   "source": [
    "# define the save path\n",
    "cwd = os.getcwd()\n",
    "save_path = os.path.join(os.path.dirname(cwd), \"workshop_5_plate_reader_output\")\n",
    "\n",
    "# create the save path, if it does not exist\n",
    "os.makedirs(save_path, exist_ok = True)\n",
    "\n",
    "# create plate reader with this folder\n",
    "reader = PlateReader(save_folder_path = save_path)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c2a790ea",
   "metadata": {},
   "source": [
    "Now, let's create a 96-well plate as an example."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8b7ebf04",
   "metadata": {},
   "outputs": [],
   "source": [
    "# We'll demonstrate functionality with both a 96 and 384 well plate\n",
    "plate = corning_96_wellplate_360ul_flat(\"plate\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c228c6a5",
   "metadata": {},
   "source": [
    "Since the peripherals aren't implemented in the visualizer yet, we won't be able to see the movement of the plates to the PlateReader. That's okay, though. We can use our imaginations and pass in the plate to our reader.\n",
    "\n",
    "Let's simulate reading this plate with our plate reader."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fda218e1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# d the plate and save the data to file\n",
    "reader.read_plate(\n",
    "    plate = plate,\n",
    "    filename = \"96_well_plate.csv\"\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "652c0451",
   "metadata": {},
   "source": [
    "One of the key ideas of using the PlateReader is that it will only give us a raw signal of absorbance. This is **NOT** the same as concentration.\n",
    "\n",
    "Let's look at our absorbance values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "531d47d3",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"Raw Absorbance\")\n",
    "df_path = os.path.join(reader.save_folder, \"96_well_plate.csv\")\n",
    "pd.read_csv(df_path,\n",
    "            index_col = 0)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2e40214b",
   "metadata": {},
   "source": [
    "In order to convert these **absorbance** values to **concentration**, we should make use of the Beer-Lambert law:\n",
    "\n",
    "According to the Beer-Lambert Law, the concentration of a compound has a linear correlation to its absorbance at a defined wavelength at a constant pathlength:\n",
    "\n",
    "$$ A = \\epsilon \\cdot C \\cdot d$$\n",
    "\n",
    "where:\n",
    "\n",
    "- $A$ = Abosrbance\n",
    "\n",
    "- $\\epsilon$ = Extinction (Molar Absorption) Coefficient\n",
    "\n",
    "- $C$ = Concentration\n",
    "\n",
    "- $d$ = Pathlength (in cm)\n",
    "\n",
    "When measuring, e.g., nucleic acid (DNA, RNA) absorbance on a 96-well plate, the liquid pathlength is not fixed to 1 cm, so the absorbance values cannot be used as\n",
    "such for concentration calculation. Therefore, the true liquid pathlength of each well must be known before the BeerLambert Law can be applied.\n",
    "\n",
    "One easy solution to this is to calculate the concentration is to simply use a series of **concentration standards**, where several samples of known concentration are used to create a line of best fit. This method avoids path-length correction and enables conversion of abosrbance to concentration.\n",
    "\n",
    "**Note -** If you would like to read more, see this [guide](https://documents.thermofisher.com/TFS-Assets/LCD/Application-Notes/AN-SkanIT-Microplate-Based-Pathlength-Correction-Technical-Note-EN.pdf) by Thermo Scientific\n",
    "\n",
    "For now, let's simulate creating this line of best fit with a function"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fd0ccce7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Linear equation parameters (example values)\n",
    "slope = 100.0  # Example slope\n",
    "intercept = 0.0  # Example intercept\n",
    "\n",
    "# function to simulate and calculate the concentration of a sample\n",
    "def read_and_calculate_concentrations(file_path: str | os.PathLike,\n",
    "                                      slope: float,\n",
    "                                      intercept: float):\n",
    "    # read data\n",
    "    df = pd.read_csv(file_path, index_col=0)\n",
    "\n",
    "    # for each row and column\n",
    "    for row in df.index:\n",
    "        for col in df.columns:\n",
    "\n",
    "            # convert the absorbance to concentration\n",
    "            absorbance = df.at[row, col]\n",
    "            concentration = absorbance * slope + intercept\n",
    "            df.at[row, col] = concentration\n",
    "            \n",
    "    return df\n",
    "\n",
    "# Process the data\n",
    "conc_df = read_and_calculate_concentrations(df_path, slope, intercept)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6b8e5146",
   "metadata": {},
   "source": [
    "Now we have a dataframe of the concentration of sample in each well. Let's visualize this"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a53c49d0",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"Concentration of Sample (ng/uL)\")\n",
    "conc_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e111709a",
   "metadata": {},
   "source": [
    "We now have a dataframe of the concentrations of our sample in each well. It is a common operation to take a plate of randomly distributed samples and do a dilution to make everything a uniform concentration.\n",
    "\n",
    "Recall the dilution equation:\n",
    "\n",
    "$$C_1V_1 = C_2V_2$$\n",
    "\n",
    "where:\n",
    "\n",
    "- $C_1$ - Concentration of solution 1.\n",
    "\n",
    "- $V_1$ - Volume of solution 1.\n",
    "\n",
    "- $C_2$ - Concentration of solution 2.\n",
    "\n",
    "- $V_2$ - Volume of solution 2.\n",
    "\n",
    "Let's create a function that can implement this calculation for our plate."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a33c9d02",
   "metadata": {},
   "outputs": [],
   "source": [
    "# C1 * V1 = C2 * V2\n",
    "def calculate_water_volume(target_conc: float,\n",
    "                           initial_volume: float,\n",
    "                           initial_conc: float):\n",
    "\n",
    "    # calculate final volume needed\n",
    "    final_volume = (initial_conc * initial_volume) / target_conc\n",
    "\n",
    "    # calculate volume to add to read final_volume\n",
    "    volume_of_diluent = final_volume - initial_volume\n",
    "\n",
    "    return volume_of_diluent\n",
    "\n",
    "# Method to convert the original dataframe into the amount of water needed to dilute each well to a target concentraiton.\n",
    "def apply_dilution(conc_df: pd.DataFrame,\n",
    "                   target_conc: float,\n",
    "                   initial_volume: float):\n",
    "\n",
    "    # copy the dataframe so we don't overwrite the original\n",
    "    dilution_df = conc_df.copy()\n",
    "\n",
    "    # for each row and col\n",
    "    for row in conc_df.index:\n",
    "        for col in conc_df.columns:\n",
    "\n",
    "            # conver the concentration to the amount dilution needd\n",
    "            initial_conc = conc_df.at[row, col]\n",
    "            volume_of_water = calculate_water_volume(target_conc, initial_volume, initial_conc)\n",
    "            dilution_df.at[row, col] = volume_of_water\n",
    "\n",
    "    return dilution_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2a9c9d25",
   "metadata": {},
   "source": [
    "Now let's run and print the results of this function"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "71f35764",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Target final concentration and initial volume\n",
    "target_conc = 20  # ng/uL\n",
    "initial_volume = 50  # uL\n",
    "\n",
    "water_to_add = apply_dilution(conc_df, target_conc, initial_volume)\n",
    "print(\"Volume of water to add (uL)\")\n",
    "water_to_add"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "241358c7",
   "metadata": {},
   "source": [
    "If we change the target concentraiton to a higher value, let's see what happens to the volume to add."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "650cc005",
   "metadata": {},
   "outputs": [],
   "source": [
    "target_conc = 50  # ng/uL\n",
    "initial_volume = 50  # uL\n",
    "\n",
    "water_to_add = apply_dilution(conc_df, target_conc, initial_volume)\n",
    "print(\"Volume of water to add (uL)\")\n",
    "water_to_add"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b8577e2c",
   "metadata": {},
   "source": [
    "Notice that some of these volumes to add are actually **negative**, which means there is not enough **protein** itself to reach the desired concentration.\n",
    "\n",
    "Since you can't easily remove the solvent from the wells (as it will contain the protein of interest) by pipetting, you should consider that when making a liquid handling protocol."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2cd4f6ca",
   "metadata": {},
   "source": [
    "### Cell Counter\n",
    "\n",
    "While tabular data is created by a plate reader, sometimes you may have interest in simulating **image data** such as microscopy images of cell cultures, which is critical to estimating cell confluency. A good introduction to the importance of cell culture comes from [this article](https://www.thermofisher.com/blog/life-in-the-lab/how-to-measure-cell-confluency/#h-what-is-cell-confluency-nbsp) from ThermoFisher:\n",
    "\n",
    ">**Cell confluency** is a routine measurement used to track cell proliferation during cell culture. It is not the absolute cell number, but rather the percentage of culture dish or flask area covered by adherent cells.\n",
    ">\n",
    ">Cell confluency is a crucial parameter that helps researchers determine timing for passaging, transfecting, or harvesting for downstream applications like drug treatments, cell therapies, induced pluripotent stem cell (iPSC) work, differentiation experiments, and more. Essentially, it’s a non-negotiable step in maintaining healthy cell cultures and ensuring accurate experimental results.\n",
    ">\n",
    ">Overconfluency in particular can lead to cell stress or death as culture nutrients deplete, and cells begin competing for physical space. Cell behavior in a crowded environment may misrepresent natural gene expression, growth, or morphologies. Lysing cells will also release cytotoxin debris, and overconfluent cell cultures in general are more susceptible to fungal and bacterial contamination – issues that can irretrievably ruin your culture.\n",
    "\n",
    "Let's create a method for **simulating images** for cell culture microscopy. Let's define a `CellImager`\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4e57f57f",
   "metadata": {},
   "outputs": [],
   "source": [
    "class CellImager:\n",
    "\n",
    "    # constructor to create the imager\n",
    "    def __init__(self,\n",
    "                 image_output_dir: str | os.PathLike):\n",
    "\n",
    "        # we need to point the cell imager to a directory since it will be saving many files\n",
    "        self.image_output_dir = image_output_dir\n",
    "\n",
    "    # method to generate cell images\n",
    "    def _generate_cell_image(self,\n",
    "                            confluency: float):\n",
    "        \"\"\"\n",
    "        Creates a single synthetic image by iteratively drawing until the target confluency is reached.\n",
    "        \"\"\"\n",
    "        \n",
    "        # create gray background 256x256 image\n",
    "        img = Image.new('RGB', [256, 256], color = \"lightgrey\")\n",
    "        draw = ImageDraw.Draw(img)\n",
    "\n",
    "        # target confluency as total percent of pixels\n",
    "        total_pixels = 256 * 256\n",
    "        target_pixel_count = total_pixels * (confluency / 100.0)\n",
    "        \n",
    "        # draw circular cells until max attempts or target pixels are covered\n",
    "        current_cell_pixels = 0\n",
    "        max_attempts = int(target_pixel_count * 5) # Safety break\n",
    "        attempts = 0\n",
    "\n",
    "        while current_cell_pixels < target_pixel_count and attempts < max_attempts:\n",
    "\n",
    "            # get random x, y coords\n",
    "            x = np.random.randint(0, 256)\n",
    "            y = np.random.randint(0, 256)\n",
    "            \n",
    "            # if the pixel is still the background color\n",
    "            if img.getpixel((x, y)) == (211, 211, 211):\n",
    "\n",
    "                # draw the circle\n",
    "                radius = np.random.randint(2, 5)\n",
    "                draw.ellipse((x - radius, y - radius, x + radius, y + radius), fill=\"red\")\n",
    "                current_cell_pixels += np.pi * radius**2\n",
    "            \n",
    "            # increment attempts\n",
    "            attempts += 1\n",
    "            \n",
    "        return img\n",
    "\n",
    "    # given an input plate, capture an image for each well\n",
    "    def capture_images_for_plate(self,\n",
    "                                 plate: Plate):\n",
    "        \"\"\"\n",
    "        Function to actually capture image for a plate\n",
    "        \"\"\"\n",
    "\n",
    "        # create folder to store images\n",
    "        if not os.path.exists(self.image_output_dir):\n",
    "            os.makedirs(self.image_output_dir)\n",
    "            print(f\"Created directory: {self.image_output_dir}\")\n",
    "\n",
    "        # capture images for each well\n",
    "        image_paths = {}\n",
    "        for well in plate.get_all_items():\n",
    "\n",
    "            # get a random confluence value between 10 - 120%\n",
    "            random_confluency = np.random.uniform(low = 10.0,\n",
    "                                                  high = 120.0)\n",
    "            \n",
    "            # generate an image for that well\n",
    "            img = self._generate_cell_image(confluency=random_confluency)\n",
    "\n",
    "            # write the image to file\n",
    "            file_path = os.path.join(self.image_output_dir, f\"well_{well.name}.png\")\n",
    "            img.save(file_path)\n",
    "            \n",
    "            # update the returned mapping to contain the file for the well in question\n",
    "            image_paths[well.name] = file_path\n",
    "            \n",
    "        print(f\"Finished capturing {len(image_paths)} images.\")\n",
    "        return image_paths"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bc2004d8",
   "metadata": {},
   "source": [
    "Now let's instantiate the image capturer."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0cbc6152",
   "metadata": {},
   "outputs": [],
   "source": [
    "# make path for imager results\n",
    "cwd = os.getcwd()\n",
    "save_path = os.path.join(os.path.dirname(cwd), \"workshop_5_imager_data\")\n",
    "\n",
    "# create the imager\n",
    "imager = CellImager(image_output_dir = save_path)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bf910a44",
   "metadata": {},
   "source": [
    "Let's \"image' a cell culture plate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e14fdc08",
   "metadata": {},
   "outputs": [],
   "source": [
    "# call the image function to capture the images.\n",
    "image_paths = imager.capture_images_for_plate(plate)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b5083909",
   "metadata": {},
   "source": [
    "Let's visualize some of these images that the plate reader has captured."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "91c3ad5f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# these imports allow us to display images in markdown output (jupyter notebook)\n",
    "from IPython.display import display, Markdown\n",
    "from IPython.display import Image as IMG\n",
    "\n",
    "# function to display images\n",
    "def display_images(image_dir):\n",
    "\n",
    "    # for each image in the target dir\n",
    "    png_files = [f for f in os.listdir(image_dir) if f.endswith('.png')]\n",
    "\n",
    "    # show the first 3\n",
    "    for i, filename in enumerate(png_files[:3]):\n",
    "        file_path = os.path.join(image_dir, filename)\n",
    "        display(Markdown(f\"#### {filename}\"))\n",
    "        display(IMG(filename=file_path))\n",
    "        if i < 2:\n",
    "            display(Markdown(\"---\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6afe8a8d",
   "metadata": {},
   "source": [
    "Run the function. You should see several simulated \"cell images\" where the cells are red."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b42a6156",
   "metadata": {},
   "outputs": [],
   "source": [
    "display_images(imager.image_output_dir)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ecdfcf5d",
   "metadata": {},
   "source": [
    "Great. Now let's create a function that can estimate the cell confluency, given an image."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7eb91b78",
   "metadata": {},
   "outputs": [],
   "source": [
    "def analyze_confluency_from_image(file_path: str | os.PathLike,\n",
    "                                  bg_color_rgb: tuple[int] = (211, 211, 211)):\n",
    "\n",
    "    # open the image\n",
    "    img = Image.open(file_path).convert('RGB')\n",
    "    \n",
    "    # convert to a numpy array\n",
    "    img_array = np.array(img)\n",
    "    \n",
    "    # create a boolean mask where True means the pixel is NOT the background color\n",
    "    is_cell_mask = ~np.all(img_array == bg_color_rgb, axis=-1)\n",
    "    \n",
    "    # calculate confluency by counting the is_cell_mask / the total pixels\n",
    "    cell_pixel_count = np.sum(is_cell_mask)\n",
    "    total_pixel_count = img.width * img.height\n",
    "    confluency = (cell_pixel_count / total_pixel_count) * 100\n",
    "    \n",
    "    # round result to two decimal points\n",
    "    return round(confluency, 2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "142e1eb8",
   "metadata": {},
   "source": [
    "Great. Now remember our cell imager returned to us the saved location of each well and its name in the form of a dictionary. Let's look at some of the items in this dictionary:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "68afc017",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(image_paths.keys()) # keys are the well names\n",
    "print(image_paths.values()) # the values are the file paths to the image of the well in the key"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "23b1c43e",
   "metadata": {},
   "source": [
    "Let's write a **for** loop that can iterate over these items, and for each well calculate a binary yes or no value if that cell is ready for passaging."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "21453ece",
   "metadata": {},
   "outputs": [],
   "source": [
    "def predict_cell_confluency(image_paths: dict, confluency_threshold: float):\n",
    "\n",
    "    # for each well and image location in our returned data\n",
    "    for well_name, image_path in image_paths.items():\n",
    "\n",
    "        # calcualte it's confluency\n",
    "        measured_confluency = analyze_confluency_from_image(image_path)\n",
    "\n",
    "        # using a threshold value of our choice, print whether or not that cell is ready for passaging.\n",
    "        print(f\"Well {well_name}: Measured Confluency = {measured_confluency}%\")\n",
    "        if measured_confluency > confluency_threshold:\n",
    "            print(f\"  -> Decision: Well {well_name} is ready for passaging.\")\n",
    "        else:\n",
    "            print(f\"  -> Decision: Well {well_name} needs more incubation time.\")\n",
    "\n",
    "predict_cell_confluency(image_paths, 80.00)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8e616ba0",
   "metadata": {},
   "source": [
    "## Exercises\n",
    "\n",
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "06773795",
   "metadata": {},
   "source": [
    "### **Exercise 1.** Using Plate Reader (60 pts)\n",
    "\n",
    "**1.A.** Implement plate reader\n",
    "\n",
    "Our plate reader in the prior problem was hard coded to only handle 96-well plates. Fill in the methods below to create a `GeneralPlateReader` class which can operate on any size plate.\n",
    "\n",
    "You should implement\n",
    "\n",
    "- `read_plate` - Input a plate and parameters for an absorbance distribution and create a `self.df` attribute with the return dataframe.\n",
    "\n",
    "- `calculate_concentrations` - Uses the `self.df` attribute, and inputs a slope and intercept for a simulated line of best fit from a standard curve to update the `self.df` attribute to contain the values in concentrations.\n",
    "\n",
    "- `calculate_water_volume` - Inputs a target concentration, initial volume, and initial concentration to calculate the water needed to add to a given well.\n",
    "\n",
    "- `apply_dilutions` - Uses the `calculate_water_volume` method and the `self.df` attribute to convert all the concentrations of `self.df` into the volume needed to add to each well.\n",
    "\n",
    "**Note -** Comment your code and submit it as `exercise_1A.txt`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e035797b",
   "metadata": {},
   "outputs": [],
   "source": [
    "class GeneralPlateReader:\n",
    "\n",
    "    # define the constructor which can take in a place to save the absorbance data\n",
    "    def __init__(self,\n",
    "                 save_folder_path: str | os.PathLike):\n",
    "        self.save_folder = save_folder_path\n",
    "\n",
    "\n",
    "    # create a function to simulate reading a 96-well plate and saving the absorbance values to file.\n",
    "    def read_plate(self,\n",
    "                   plate: Plate,\n",
    "                   mean_abs: float,\n",
    "                   abs_stdev: float):\n",
    "        \n",
    "        ... # YOUR CODE HERE\n",
    "        # assign output as df attribute of class\n",
    "        self.df = df\n",
    "\n",
    "    def calculate_concentrations(self,\n",
    "                                 slope: float,\n",
    "                                 intercept: float):\n",
    "\n",
    "        ... # YOUR CODE HERE\n",
    "\n",
    "    def calculate_water_volume(self,\n",
    "                               target_conc: float,\n",
    "                               initial_volume: float,\n",
    "                               initial_conc: float):\n",
    "\n",
    "        ... # YOUR CODE HERE\n",
    "        return volume_of_diluent\n",
    "\n",
    "    # Method to convert the original dataframe into the amount of water needed to dilute each well to a target concentraiton.\n",
    "    def apply_dilution(self,\n",
    "                       target_conc: float,\n",
    "                       initial_volume: float):\n",
    "\n",
    "        ... # YOUR CODE HERE\n",
    "\n",
    "# instantiate plate reader and example plate\n",
    "reader = GeneralPlateReader(save_folder_path = save_path)\n",
    "plate = corning_96_wellplate_360ul_flat(\"plate\")\n",
    "\n",
    "# read the plate and print results\n",
    "reader.read_plate(plate, 1, 0.5)\n",
    "print(f\"---PLATE READING RESULTS---\")\n",
    "print(reader.df)\n",
    "print()\n",
    "\n",
    "# calculate the concentrations and print results\n",
    "reader.calculate_concentrations(100, 0)\n",
    "print(f\"---CONCENTRATION CALCULATION RESULTS---\")\n",
    "print(reader.df)\n",
    "print()\n",
    "\n",
    "# apply the dilutions and print results\n",
    "reader.apply_dilution(40, 20)\n",
    "print(f\"---DILUTION CALCULATION RESULTS---\")\n",
    "print(reader.df)\n",
    "print()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0e195fb2",
   "metadata": {},
   "source": [
    "**1.B.** 20 ng / uL Dilutions\n",
    "\n",
    "Using your `GeneralPlateReader` class, you will need to implement a liquid handling protocol which starts with a DNA plate, tips, and a reservoir of water, reads the plate, and then using the calculated volumes to pipette, performs the dilutions. Your code should contain:\n",
    "\n",
    "- Relevant imports to execute your script\n",
    "\n",
    "- A `make_deck()` function which inputs:\n",
    "\n",
    "    - A `plate_type` to create either a 96 well or 24 well plate for dilutions\n",
    "\n",
    "    - An `initial_vol` argument to set the initial volume of DNA on the DNA plate.\n",
    "\n",
    "    - The function should then:\n",
    "\n",
    "        - Create an OpenTrons Deck\n",
    "\n",
    "        - Assign the DNA plate to the deck\n",
    "\n",
    "        - Assign tips to the deck\n",
    "\n",
    "        - Set the initial DNA liquid levels\n",
    "\n",
    "        - Set the initial dilution reservoir level of water to 90 mL\n",
    "\n",
    "- A `remove_plate_and_read()` function that simulates a gripper moving your plate to a plate reader and returning data. We have already implemented this function for you.\n",
    "\n",
    "- A helper function to perform a **mixing** operation and a helper function to create the **tip generator**.\n",
    "\n",
    "- A `pipette_vols()` function that inputs a row and column to move diluent from the diluent reservoir to the DNA plate at the row-column position.\n",
    "\n",
    "Using these functions, write a liquid handling protocol that:\n",
    "\n",
    "1. Set up the OT-2 Deck\n",
    "\n",
    "2. Reads the plate using the defined function\n",
    "\n",
    "3. Pipettes the correct volume to dilute each sample to 20 ng / uL\n",
    "\n",
    "Make sure to **record a GIF** including the deck setup and pipetting operations. Use the following settings for plate reading:\n",
    "\n",
    "- mean_abs = 0.5\n",
    "\n",
    "- abs_stdev = 0.1\n",
    "\n",
    "- slope = 100\n",
    "\n",
    "- intercept = 0\n",
    "\n",
    "- target_conc = 20\n",
    "\n",
    "- initial_volume = 50\n",
    "\n",
    "**Note -** Comment your code and submit it as `exercise_1B.txt` and `exercise_1B.gif`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "453fbf9a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Imports\n",
    "import time\n",
    "from pylabrobot.resources import Resource\n",
    "import ... # YOUR CODE HERE\n",
    "\n",
    "def make_deck(plate_type = \"96_well\", initial_vol = 50):\n",
    "\n",
    "    if plate_type == \"96_well\":\n",
    "        dna_plate = ... # YOUR CODE HERE\n",
    "    elif plate_type == \"24_well\":\n",
    "        dna_plate = ... # YOUR CODE HERE\n",
    "\n",
    "    ... # YOUR CODE HERE\n",
    "    return deck\n",
    "\n",
    "def remove_plate_and_read(lh,\n",
    "                          plate_name: str,\n",
    "                          plate_slot: int,\n",
    "                          mean_abs: float, \n",
    "                          abs_stdev: float,\n",
    "                          slope: float, \n",
    "                          intercept: float,\n",
    "                          target_conc: float,\n",
    "                          initial_volume: float):\n",
    "\n",
    "    # get plate and \"move it\" to plate reader\n",
    "    plate = lh.get_resource(plate_name)\n",
    "    lh.deck.unassign_child_resource(plate)\n",
    "\n",
    "    # read the plate\n",
    "    reader.read_plate(plate, mean_abs, abs_stdev)\n",
    "    time.sleep(5) # simulate plate reading time\n",
    "\n",
    "    # calculate the concentrations\n",
    "    reader.calculate_concentrations(slope, intercept)\n",
    "\n",
    "    # calculat the amount of water to add\n",
    "    reader.apply_dilution(target_conc, initial_volume)\n",
    "\n",
    "    # put the plate back onto the deck\n",
    "    lh.deck.assign_child_at_slot(plate, plate_slot)\n",
    "    return reader.df\n",
    "\n",
    "def prepare_tip_generator(\n",
    "        tip_racks\n",
    "):\n",
    "    ... # YOUR CODE HERE\n",
    "\n",
    "async def mix(\n",
    "        lh,\n",
    "        plate,\n",
    "        well,\n",
    "        n_mixes,\n",
    "        vol\n",
    "):\n",
    "    ... # YOUR CODE HERE\n",
    "\n",
    "async def pipette_vols(lh: LiquidHandler,\n",
    "                 diluent_reservoir: Resource,\n",
    "                 dna_plate: Plate,\n",
    "                 tip_generator,\n",
    "                 row: str,\n",
    "                 col: str,\n",
    "                 volume: float):\n",
    "\n",
    "    ... # YOUR CODE HERE\n",
    "\n",
    "async def run_protocol_1B(lh):\n",
    "\n",
    "    ... # YOUR CODE HERE\n",
    "    vol_df = remove_plate_and_read(...)\n",
    "\n",
    "    # for each row and col, pipette the volumes into the dna plate\n",
    "    ... # YOUR CODE HERE\n",
    "\n",
    "# run the protocol\n",
    "deck = make_deck()\n",
    "lh = await visualize_deck(deck, LiquidHandlerChatterboxBackend())\n",
    "await run_protocol_1B(lh)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9ebc35cd",
   "metadata": {},
   "source": [
    "**1.C.** 40 ng / uL diliutions\n",
    "\n",
    "Repeat the dilution experiment above, but this time, implement a function called `pipette_vols_careful()`, which implements a logical check to skip the well if:\n",
    "\n",
    "- The volume to pipette is negative, or...\n",
    "\n",
    "- The volume is greater than the biggest tip capacity, or...\n",
    "\n",
    "- The volume to pipette would cause the current volume in the well to exceed its maximum capacity.\n",
    "\n",
    "Using these functions, write a liquid handling protocol that:\n",
    "\n",
    "1. Set up the OT-2 Deck\n",
    "\n",
    "2. Reads the plate using the defined function\n",
    "\n",
    "3. Pipettes the correct volume to dilute each sample to 40 ng / uL\n",
    "\n",
    "Make sure to **record a GIF** including the deck setup and pipetting operations. Use the following settings for plate reading:\n",
    "\n",
    "- mean_abs = 0.5\n",
    "\n",
    "- abs_stdev = 0.1\n",
    "\n",
    "- slope = 100\n",
    "\n",
    "- intercept = 0\n",
    "\n",
    "- target_conc = 40\n",
    "\n",
    "- initial_volume = 2000\n",
    "\n",
    "**Note -** You may have to use larger tips (i.e. adjust your make_deck function). Comment your code and submit it as `exercise_1C.txt` and `exercise_1C.gif`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cbaa93ed",
   "metadata": {},
   "outputs": [],
   "source": [
    "async def pipette_vols_careful(lh: LiquidHandler,\n",
    "                 diluent_reservoir: Resource,\n",
    "                 dna_plate: Plate,\n",
    "                 tip_generator,\n",
    "                 row: str,\n",
    "                 col: str,\n",
    "                 volume: float):\n",
    "\n",
    "    ... # YOUR CODE HERE\n",
    "\n",
    "async def run_protocol_1C(lh):\n",
    "\n",
    "    ... # YOUR CODE HERE\n",
    "    vol_df = remove_plate_and_read(...)\n",
    "\n",
    "    ... # YOUR CODE HERE\n",
    "\n",
    "# run the protocol\n",
    "deck = make_deck(plate_type = \"24_well\", initial_vol = 2000)\n",
    "lh = await visualize_deck(deck, LiquidHandlerChatterboxBackend())\n",
    "await run_protocol_1C(lh)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "63fc3190",
   "metadata": {},
   "source": [
    "### **Exercise 2.** Cell Counter + Passaging (40 pts)\n",
    "\n",
    "**2.A.** Modified Cell Counter\n",
    "\n",
    "We will be creating a modified version of the cell counter which is able to simulate confluency **over several days** of cell growth. Below, we provide a template `GrowthSimulatedCellImager` class with some methods defined. You will need to implement the following methods:\n",
    "\n",
    "- `__init__()`\n",
    "\n",
    "    - This function should now take in new growth simulation parameters and assign them to the `self` attribute of the class.\n",
    "\n",
    "    - Also you should create a `self.confluent_wells` attribute that is a python `set()` that can store the wells that are currently confluent.\n",
    "\n",
    "- `_simulate_confluency()`\n",
    "\n",
    "    - This method hsould take in an argument named `time_point` to simulate the base confluency with the following equation:\n",
    "\n",
    "        - $B = I \\cdot r^t$ where $I$ is the initial confluency, $r$ is the growth rate and $t$ is the time point.\n",
    "\n",
    "        - $B_{noise} = B + C_{noise}$ and $C_{noise}$ is drawn from a uniform distribution $\\left[-\\text{cell variability}, +\\text{cell variability}\\right]$\n",
    "\n",
    "        - Use [np.clip](https://numpy.org/doc/2.1/reference/generated/numpy.clip.html) to clip the value beetween $0$ and `max_confluency`\n",
    "\n",
    "- `capture_images_for_plate()`\n",
    "\n",
    "    - Same as the prior capture image function, but should check if a well is already in the `self.confluent_wells` attribute\n",
    "\n",
    "- `predict_cell_confluency()`\n",
    "\n",
    "    - For each well name and image path in the input `well_data` attribute and `confluency_threshold`,  add to the `self.confluent_wells` attribute wells above the `confluency_threshold` paramter.\n",
    "\n",
    "    - For each well ready for passaging, return it as a `list()` so your liquid handler knows which wells to passage.\n",
    "\n",
    "**Note -** Comment your code and submit it as `exercise_2A.txt`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0651a188",
   "metadata": {},
   "outputs": [],
   "source": [
    "class GrowthSimulatedCellImager:\n",
    "\n",
    "    # constructor with variables for growth rate and confluency measurements\n",
    "    def __init__(self,\n",
    "                 image_output_dir: str | os.PathLike,\n",
    "                 initial_confluency: float = 10.0,\n",
    "                 growth_rate: float = 1.6,\n",
    "                 max_confluency: float = 120.0,\n",
    "                 per_well_variability: float = 30.0\n",
    "                ):\n",
    "\n",
    "        ... # YOUR CODE HERE\n",
    "\n",
    "        # way to track current wells\n",
    "        self.confluent_wells = ... # YOUR CODE HERE\n",
    "\n",
    "    # simulate growth over time\n",
    "    def _simulate_confluency(self, time_point: int) -> float:\n",
    "\n",
    "        base_confluency = ... # YOUR CODE HERE\n",
    "        noise = np.random.uniform(...) # YOUR CODE HERE\n",
    "        simulated_confluency = base_confluency + noise # YOUR CODE HERE\n",
    "        \n",
    "        # clamp the value between 0 and the maximum confluency\n",
    "        return np.clip(simulated_confluency, 0, self.max_confluency)\n",
    "\n",
    "    def _generate_cell_image(self,\n",
    "                             confluency: float):\n",
    "        \n",
    "        # draw background image\n",
    "        img = Image.new('RGB', [256, 256], color = \"lightgrey\")\n",
    "        draw = ImageDraw.Draw(img)\n",
    "\n",
    "        # target confluency\n",
    "        total_pixels = 256 * 256\n",
    "        target_pixel_count = total_pixels * (confluency / 100.0)\n",
    "        \n",
    "        # keep drawing cells until confluency reached or break condition reached\n",
    "        current_cell_pixels = 0\n",
    "        max_attempts = int(target_pixel_count * 5) + 500 # safety break\n",
    "        attempts = 0\n",
    "\n",
    "        # draw cells until confluency reached\n",
    "        while current_cell_pixels < target_pixel_count and attempts < max_attempts:\n",
    "\n",
    "            # get random x, y coords\n",
    "            x = np.random.randint(0, 256)\n",
    "            y = np.random.randint(0, 256)\n",
    "            \n",
    "            # if the pixel is still the background color\n",
    "            if img.getpixel((x, y)) == (211, 211, 211):\n",
    "\n",
    "                # draw the circle\n",
    "                radius = np.random.randint(2, 5)\n",
    "                draw.ellipse((x - radius, y - radius, x + radius, y + radius), fill=\"red\")\n",
    "                current_cell_pixels += np.pi * radius**2\n",
    "            \n",
    "            # increment attempts\n",
    "            attempts += 1\n",
    "            \n",
    "        return img\n",
    "\n",
    "    def capture_images_for_plate(self,\n",
    "                                 plate: Plate,\n",
    "                                 time_point: int\n",
    "                                ):\n",
    "\n",
    "        # create folder to store images\n",
    "        if not os.path.exists(self.image_output_dir):\n",
    "            os.makedirs(self.image_output_dir)\n",
    "            print(f\"Created directory: {self.image_output_dir}\")\n",
    "\n",
    "        # dictionary to hold data\n",
    "        well_data_from_run = {}\n",
    "        \n",
    "        # capture images for each well\n",
    "        for well in plate.get_all_items():\n",
    "\n",
    "            # if the well name is already confluent, skip the image\n",
    "            well_name = well.name.split(\"_\")[-1]\n",
    "            if ... # YOUR CODE HERE\n",
    "\n",
    "            # otherwise, image the cell\n",
    "            simulated_confluency = self._simulate_confluency(time_point=time_point)\n",
    "            img = self._generate_cell_image(confluency=simulated_confluency)\n",
    "            file_path = os.path.join(self.image_output_dir, f\"well_{well_name}_t{time_point}.png\")\n",
    "            img.save(file_path)\n",
    "\n",
    "            # save the image path\n",
    "            well_data_from_run[well_name] = file_path\n",
    "            \n",
    "        return well_data_from_run\n",
    "\n",
    "    def analyze_confluency_from_image(self,\n",
    "                                      file_path: str | os.PathLike,\n",
    "                                      bg_color_rgb: tuple[int] = (211, 211, 211)):\n",
    "        \"\"\"\n",
    "        Analyzes the actual pixel-based confluency from a saved image.\n",
    "        (This function is unchanged)\n",
    "        \"\"\"\n",
    "        # open the image\n",
    "        img = Image.open(file_path).convert('RGB')\n",
    "        \n",
    "        # convert to a numpy array\n",
    "        img_array = np.array(img)\n",
    "        \n",
    "        # Create a boolean mask where True means the pixel is NOT the background color\n",
    "        is_cell_mask = ~np.all(img_array == bg_color_rgb, axis=-1)\n",
    "        \n",
    "        # Calculate confluency\n",
    "        cell_pixel_count = np.sum(is_cell_mask)\n",
    "        total_pixel_count = img.width * img.height\n",
    "        confluency = (cell_pixel_count / total_pixel_count) * 100\n",
    "        \n",
    "        return round(confluency, 2)\n",
    "\n",
    "    def predict_cell_confluency(self,\n",
    "                                well_data,\n",
    "                                confluency_threshold: float\n",
    "                               ):\n",
    "\n",
    "        ready_for_passaging = []\n",
    "        \n",
    "        print(f\"\\n--- Confluency Analysis (Threshold: {confluency_threshold}%) ---\")\n",
    "\n",
    "        # for each well in the dictionary\n",
    "        for well_name, image_path in well_data.items():\n",
    "            ... # YOUR CODE HERE\n",
    "            \n",
    "            # if confluent, add to confluent wells\n",
    "            if ... # YOUR CODE HERE\n",
    "                print(f\"  -> Decision: Well {well_name} is ready for passaging.\")\n",
    "                self.confluent_wells.add(...) # YOUR CODE HERE\n",
    "                ready_for_passaging.append(...) # YOUR CODE HERE\n",
    "            \n",
    "            # Otherwise, skip\n",
    "            else:\n",
    "                print(f\"  -> Decision: Well {well_name} needs more incubation time.\")\n",
    "                \n",
    "        # return list of wells ready for passaging\n",
    "        return ..."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "93fb3b5b",
   "metadata": {},
   "source": [
    "**2.B.** Cell Passaging Liquid Handling\n",
    "\n",
    "Using your `GrowthSimulatedCellImager` class, you will need to implement a liquid handling protocol which starts with a cell culture plate, tips, and reservoirs of PBS, trypsin, and media.\n",
    "\n",
    "The protocol should then simulate **5 days** of cell growth, imaging the wells on the plate that are still growing each day, and then passaging any wells that need passaging. Your script should include:\n",
    "\n",
    "- Relevant imports to execute your script\n",
    "\n",
    "- A `make_deck()` function which creates a deck with:\n",
    "\n",
    "    - An OpenTrons Deck\n",
    "\n",
    "    - A cell culture plate\n",
    "\n",
    "    - A destination plate\n",
    "\n",
    "    - Enough tip racks (may need 7 or 8 here)\n",
    "\n",
    "    - A 6-well reservoir with enough PBS, Media, and Trypsin to complete the protocol\n",
    "\n",
    "- A `prepare_new_plate_media()` function which pipettes fresh starting media (180 uL per well) from the media reservoir to the destination plate.\n",
    "\n",
    "- A `passage_well()` function that inputs the well name to passage along with other parameters and:\n",
    "\n",
    "    - Aspirates all media (200 uL) from the well\n",
    "\n",
    "    - Washes the well with 100 uL of PBS\n",
    "\n",
    "    - Detaches the cells with 50 uL of Trypsin\n",
    "\n",
    "    - Inactivates the trypsin with 50 uL of cell media\n",
    "\n",
    "    - Aspirates the trypsin-cell-media mixture and moves it to the corresponding well on the destination plate at a 1:10 ratio (20 uL in 180 uL)\n",
    "\n",
    "- A `simulate_imaging()` function that simulates a gripper moving your plate to a cell imager, and returns a dictionary of well_id to image path data of the wells imaged (or returns None if all wells imaged are confluent). We implement this for you.\n",
    "\n",
    "- A helper function to perform a **mixing** operation and a helper function to create the **tip generator**.\n",
    "\n",
    "Using these functions, write a liquid handling protocol called `run_protocol_2B()` that:\n",
    "\n",
    "- Get's all relevant resources from the deck.\n",
    "\n",
    "- Instantiates a `GrowthSimulatedCellImager` class.\n",
    "\n",
    "- Prepares the destination plate with fresh media in all wells.\n",
    "\n",
    "- For each day in 5 days, the protocol should:\n",
    "\n",
    "    - Call the `simulate_imaging` function to get the wells images for wells not passaged so far.\n",
    "\n",
    "    - Using these images, call `predict_cell_confluency()` on your imager to get the wells to passage, if any.\n",
    "\n",
    "    - For each well in the wells to passage, call your `passage_well()` function to passage that well to the new destination plate at a 1:10 ratio.\n",
    "\n",
    "    - If no new wells exist in the `simulate_imaging()` function, break the loop early.\n",
    "\n",
    "    - Use the default class parameters we define in **Exercise 2A.** Use **80%** as the confluence threshold for passaging.\n",
    "\n",
    "**Note -** Comment your code and submit it as `exercise_2B.txt` and `exercise_2B.gif`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d0899f5c",
   "metadata": {},
   "outputs": [],
   "source": [
    "from pylabrobot.resources import Porvair_6_reservoir_47ml_Vb\n",
    "\n",
    "def make_deck_2B():\n",
    "    ... # YOUR CODE HERE\n",
    "    return deck\n",
    "\n",
    "def prepare_tip_generator(\n",
    "        tip_racks\n",
    "):\n",
    "    ... # YOUR CODE HERE\n",
    "\n",
    "async def mix(\n",
    "        lh,\n",
    "        plate,\n",
    "        well,\n",
    "        n_mixes,\n",
    "        vol\n",
    "):\n",
    "    ... # YOUR CODE HERE\n",
    "\n",
    "async def prepare_new_plate_media(lh: LiquidHandler,\n",
    "                                  media_reservoir: Resource,\n",
    "                                  media_well: str,\n",
    "                                  target_plate: Plate,\n",
    "                                  volume: float,\n",
    "                                  tip_gen):\n",
    "    ... # YOUR CODE HERE\n",
    "\n",
    "async def passage_well(lh: LiquidHandler,\n",
    "                       well_name: str,\n",
    "                       tip_gen,\n",
    "                       reservoir_plate: Plate,\n",
    "                       media_well: str,\n",
    "                       pbs_well: str,\n",
    "                       trypsin_well: str,\n",
    "                       cell_plate: Plate,\n",
    "                       destination_plate: Plate):\n",
    "\n",
    "    # 1. Aspirate Media\n",
    "    ... # YOUR CODE HERE\n",
    "\n",
    "    # 2. Wash with PBS\n",
    "    ... # YOUR CODE HERE\n",
    "\n",
    "    # 3. Detach with Trypsin\n",
    "    ... # YOUR CODE HERE\n",
    "\n",
    "    # 4. Neutralize and 5. Transfer (split)\n",
    "    ... # YOUR CODE HERE\n",
    "    \n",
    "    # Transfer for a 1:10 split (20uL from 100uL total, into 180uL new media)\n",
    "    ... # YOUR CODE HERE\n",
    "\n",
    "async def simulate_imaging(lh, plate_name, plate_slot, imager, day):\n",
    "\n",
    "    # get plate and \"move it\" to cell imager\n",
    "    plate = lh.get_resource(plate_name)\n",
    "    lh.deck.unassign_child_resource(plate)\n",
    "    time.sleep(5)\n",
    "\n",
    "    # capture plate images\n",
    "    new_well_data = imager.capture_images_for_plate(\n",
    "        plate=plate,\n",
    "        time_point=day\n",
    "    )\n",
    "\n",
    "    # put the plate back onto the deck\n",
    "    lh.deck.assign_child_at_slot(plate, plate_slot)\n",
    "\n",
    "    return new_well_data\n",
    "\n",
    "async def run_protocol_2B(lh, num_days=5):\n",
    "\n",
    "    ... # YOUR CODE HERE\n",
    "    imager = ... # YOUR CODE HERE\n",
    "\n",
    "    # prepare new plate\n",
    "    await prepare_new_plate_media(...)\n",
    "    \n",
    "    for day in range(1, num_days + 1):\n",
    "        new_well_data = await simulate_imaging(...)\n",
    "\n",
    "        if ... # YOUR CODE HERE\n",
    "            print(\"All wells are confluent. Experiment complete.\")\n",
    "            break\n",
    "        ... # YOUR CODE HERE\n",
    "\n",
    "deck = make_deck_2B()\n",
    "lh = await visualize_deck(deck, LiquidHandlerChatterboxBackend())\n",
    "await run_protocol_2B(lh)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a00774f1",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "#### Conclusion\n",
    "\n",
    "That's all for workshop 5! Double check that you have submitted a `.txt` file for every problem, a `.gif` file for problems 1B, 1C, and 2B. You should have submitted:\n",
    "\n",
    "- `exercise_1A.txt` with the code for exercise 1A.\n",
    "\n",
    "- `exercise_1B.txt` with the code for exercise 1B.\n",
    "\n",
    "- `exercise_1B.gif` showing the full deck setup and protocol of exercise 1B\n",
    "\n",
    "- `exercise_1C.txt` with the code for exercise 1C.\n",
    "\n",
    "- `exercise_1C.gif` showing the full deck setup and protocol of exercise 1C\n",
    "\n",
    "- `exercise_2A.txt` with the code for exercise 2A.\n",
    "\n",
    "- `exercise_2B.txt` with the code for exercise 2B.\n",
    "\n",
    "- `exercise_2B.gif` showing the full deck setup and protocol of exercise 2B\n",
    "\n",
    "**EVERY CODE BLOCK SHOULD HAVE WELL-WRITTEN COMMENTS**\n",
    "\n",
    "If you are still feeling unsure on deck setup, please **reach out to the teaching team**, contact info for whom can be found in the .`README.md` file on the [class GitHub](https://github.com/chory-lab/bme590-fall-2025)\n",
    "\n",
    "---"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "lab-automation",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
